libyul/backends/evm/Dominator.h

```diff
@@ -0,0 +1,312 @@
+				return label[_vIdx];
+			}
+			return _vIdx;
+		};
+
+		auto toIdx = [&](V const& v) { return m_vertexIndices[v]; };
+
+		// step 1
+		std::set<V> visited;
+		// predecessors(w): The set of vertices ``v`` such that (``v``, ``w``) is an edge of the graph.
+		std::vector<std::set<size_t>> predecessors(numVertices);
+		// bucket(w): a set of vertices whose semidominator is ``w``
+		// The index of the array represents the vertex's ``dfIdx``
+		std::vector<std::deque<size_t>> bucket(numVertices);
+		// idom(w): the index of the immediate dominator of ``w``
+		std::vector<size_t> idom(numVertices, std::numeric_limits<size_t>::max());
+		// The number of vertices reached during the DFS.
+		// The vertices are indexed based on this number.
+		size_t dfIdx = 0;
+		auto dfs = [&](V const& _v, auto _dfs) -> void {
+			auto [_, inserted] = visited.insert(_v);
+			if (!inserted)
+				return;
+			m_vertices[dfIdx] = _v;
+			m_vertexIndices[_v] = dfIdx;
+			semi[dfIdx] = dfIdx;
+			label[dfIdx] = dfIdx;
+			dfIdx++;
+			ForEachSuccessor{}(_v, [&](V const& w) {
+				if (semi[dfIdx] == std::numeric_limits<size_t>::max())
+				{
+					parent[dfIdx] = m_vertexIndices[_v];
+					_dfs(w, _dfs);
+				}
+				predecessors[m_vertexIndices[w]].insert(m_vertexIndices[_v]);
+			});
+		};
+		dfs(_entry, dfs);
+
+		// Process the vertices in decreasing order of the DFS number
+		for (size_t wIdx: m_vertices | ranges::views::reverse | ranges::views::transform(toIdx))
+		{
+			// step 3
+			// NOTE: this is an optimization, i.e. performing the step 3 before step 2.
+			// The goal is to process the bucket in the beginning of the loop for the vertex ``w``
+			// instead of ``parent[w]`` in the end of the loop as described in the original paper.
+			// Inverting those steps ensures that a bucket is only processed once and
+			// it does not need to be erased.
+			// The optimization proposal is available here: https://jgaa.info/accepted/2006/GeorgiadisTarjanWerneck2006.10.1.pdf pg.77
+			ranges::for_each(
+				bucket[wIdx],
+				[&](size_t vIdx)
+				{
+					size_t uIdx = eval(vIdx);
+					solAssert(uIdx <= vIdx);
+					idom[vIdx] = (semi[uIdx] < semi[vIdx]) ? uIdx : wIdx;
+				}
+			);
+
+			// step 2
+			for (size_t vIdx: predecessors[wIdx])
+			{
+				size_t uIdx = eval(vIdx);
+				solAssert(uIdx <= vIdx);
+				if (semi[uIdx] < semi[wIdx])
+					semi[wIdx] = semi[uIdx];
+			}
+			bucket[semi[wIdx]].emplace_back(wIdx);
+			link(parent[wIdx], wIdx);
+			solAssert(ancestor[wIdx] == parent[wIdx]);
+		}
+
+		// step 4
+		idom[0] = 0;
+		for (size_t wIdx: m_vertices | ranges::views::drop(1) | ranges::views::transform(toIdx))
+			if (idom[wIdx] != semi[wIdx])
+				idom[wIdx] = idom[idom[wIdx]];
+
+		return idom;
+	}
+
+private:
+	/// Build dominator tree from the immediate dominators set.
+	/// The function groups all the indices that are immediately dominated by a vertex.
+	void buildDominatorTree()
+	{
+		solAssert(!m_immediateDominators.empty());
```

Comment:
This assumption won't always hold. When there are no vertices `m_immediateDominators` will remain empty even after `findDominators()`.

Reply:
This is not the case anymore and `findDominators()` asserts that the graph has at least one vertex. The current assertion in `buildDominatorTree` is: `solAssert(m_immediateDominators.size() == m_vertices.size());` which I believe will always hold.